{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Python Assignment - Beatriz Mizoguchi"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Your Mission: Complete Python code in a Jupyter Notebook ##\n",
    "## 1. Document Dr. X's function with comments and with markdown text in your Jupyter notebook\n",
    "## 2. Write a function that translates a string of nucleotides to amino acids based on Dr. X's pseudo-code suggestion\n",
    "## 3. Write an alternative translation function\n",
    "## 4. Write a function (or 3 functions) that calculates the proportion of each of 3 amino acid categories in a sequence\n",
    "## 5. In the MAIN part of the script, call your functions from 3 (or 2) and 4 and complete the empty columns in the dataframe\n",
    "## 6. Plot a bar-chart of adult mass per species\n",
    "## 7. Plot a graph that shows the amino-acid type proportions\n",
    "## 8. Write the entire dataframe to a new CSV file\n",
    "## 9. BONUS: What other visualizations, functions or tasks would you do with this dataset? Add something interesting for fun. \n",
    "## ** Do all of this in a Jupyter notebook and push it to a GitHub repository\n",
    "## ** Be sure to cite (by providing URLs) information appropriately in your documented notebook.\n",
    "## ** Commit and push your completed work in the Jupyter notebook to your repository\n",
    "## ** Create a link to your completed Jupyter notebook using https://nbviewer.jupyter.org\n",
    "## ** Send the above link *AND* the URL to your git repository to the instructors via Slack by the end of the day on March 31.\n",
    "\n",
    "## Not all of these tasks have been covered in class and you will \n",
    "## have to use online resources to find out how to do some of these tasks.\n",
    "\n",
    "\n",
    "######################## Python Translate Script ########################\n",
    "\n",
    "## Here's the start of our Python script. Thanks for completing it for me! - Dr. X\n",
    "## IMPORTANT: install biopython so that this will work\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from Bio import SeqIO   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from Bio.Data import CodonTable  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### 1) importing SeqIO (standard Sequence Input/Ouput interface)  \n",
    "##### 2) importing codon table so we can translate the sequence intro amino acids  \n",
    "##### 3) importing _pandas_ dataset with an alias \"pd\"  "
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "#%%%%%%%%%%%%%%%#\n",
    "### FUNCTIONS ###\n",
    "#%%%%%%%%%%%%%%%#\n",
    "\n",
    "## 1 ##\n",
    "## Dr. X: this gets sequences \n",
    "## please finish documenting this function with comments, or in notebook markdown text\n",
    "## be sure to provide details on return types and arguments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def get_sequences_from_file(fasta_fn):\n",
    "    sequence_data_dict = {}\n",
    "    for record in SeqIO.parse(fasta_fn, \"fasta\"):\n",
    "        description = record.description.split()\n",
    "        species_name = description[1] + \" \" + description[2]\n",
    "        sequence_data_dict[species_name] = record.seq\n",
    "    return(sequence_data_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1) defining a function to get sequence data from fasta file   \n",
    "2) transforming the sequence data into a dictionary  \n",
    "3) using Bio.SeqIO.parse( ) function, which analyses the file in parts and returns a SeqRecord (record) object, specifying that the data type is fasta  \n",
    "4) split record description by a space with the output being the file \"description\"  \n",
    "5) combining the second and third element of description, separated by a space and  putting into \"species_name\"  \n",
    "6) indexing the species_name from the sequence data dictionary   \n",
    "7) return the sequence data dictionary"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "## 2 ##\n",
    "####### YOUR STRING-TRANSLATE FUNCTION ########\n",
    "## Write a function that translates sequences\n",
    "## All sequences start at codon position 1\n",
    "## Complete a function that does this using loop over the string of nucleotides\n",
    "## Here is  some pseudo-code and suggestions\n",
    "## feel free to change the function and variable names\n",
    "# def translate_function(string_nucleotides): \n",
    "#     mito_table = CodonTable.unambiguous_dna_by_name[\"Vertebrate Mitochondrial\"] # this should work using BioPython (be sure to check what this returns)\n",
    "#     for-loop through every 3rd position in string_nucleotides to get the codon using range subsets\n",
    "#         # IMPORTANT: if the sequence has a stop codon at the end, you should leave it off\n",
    "#         # this is how you can retrieve the amino acid: mito_table.forward_table[codon]\n",
    "#         add the aa to aa_seq_string\n",
    "#     return(aa_seq_string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "stop_codon = {\"TAG\", \"TAA\", \"TGA\"}\n",
    "def translate_dna(sequence_data_dict):\n",
    "    mito_table = CodonTable.unambiguous_dna_by_name[\"Vertebrate Mitochondrial\"]\n",
    "    for i in 'sequence_data_dict':\n",
    "        range(0,leng(sequence_data_dict)-3,3)\n",
    "    return(aa_seq_string)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Defining function to translate sequence_data_dict\n",
    "creating mito_table with codon information\n",
    "for loop to translate information at string_nucleotide with codon information using the range function - initializng at position 1 through 4 (3 codons: position 1-3 -4th element not included- range of 3 bases)\n",
    "if statement saying that if it meets a stop codon, that the loop should stop\n",
    "return the results"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "## 3 ##\n",
    "####### YOUR ALTERNATIVE FUNCTION ########\n",
    "## Is there a better way to write the translation function? (Hint: yes there is.) \n",
    "## Perhaps using available BioPython library utilities?\n",
    "## Please also write this function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "sequence_data_dict.translate(to_stop=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "syntax was obtained from \"http://biopython.org/DIST/docs/tutorial/Tutorial.html#htoc25\"  \n",
    "using translate function from BioPython"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "## 4 ##\n",
    "####### YOUR COUNT AA FUNCTION ########\n",
    "## Write a function that takes an amino-acid sequence (as a BioPython Seq or as a string) \n",
    "## and computes the proportion of the 3 different categories of amino acid\n",
    "## (you may want to write this as 3 separate functions instead of just 1, it's up to you)\n",
    "## These categories are based on the \n",
    "# def get_proportion_aa_type_function(aa_seq):\n",
    "#     charged = ['R','K','D','E']\n",
    "#     polar = ['Q','N','H','S','T','Y','C','M','W']\n",
    "#     hydrophobic = ['A','I','L','F','V','P','G']\n",
    "#     for aa in charged:\n",
    "#         # count the number of times that aa appears, add to the total for charged\n",
    "#     # repeat for polar and hydrophobic\n",
    "#     return proportion_charged, proportion_polar, proportion_hydro \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "count = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n"
     ]
    }
   ],
   "source": [
    "charged = ['R','K','D','E']\n",
    "def groupby(aa_seq_string):\n",
    "    for aa in aa_seq_string:\n",
    "        if aa is charged:\n",
    "            count +=1\n",
    "print (count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n"
     ]
    }
   ],
   "source": [
    "polar = ['Q','N','H','S','T','Y','C','M','W']\n",
    "def groupby(aa_seq_string):\n",
    "    for aa in aa_seq_string:\n",
    "        if aa is polar:\n",
    "            count +=1\n",
    "print(count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n"
     ]
    }
   ],
   "source": [
    "hydrophobic = ['A','I','L','F','V','P','G']\n",
    "def groupby(aa_seq_string):\n",
    "    for aa in aa_seq_string:\n",
    "        if aa is hydrophobic:\n",
    "            count +=1\n",
    "print(count)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "defining the function of accounting the proportion of aa in the string\n",
    "conditioning the aa into their category\n",
    "for loop with the condition of finding the aa that fit the category and counting them\n",
    "print the total of each category"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "#%%%%%%%%%%%%%%#\n",
    "###   MAIN   ###\n",
    "#%%%%%%%%%%%%%%#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "cytb_seqs = open(\"bears_cytb.fasta\") "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "importing cytochrome b sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "bear_df = pd.read_csv(\"bears_data.csv\") # Includes only data for body mass currently and empty cells for other columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "creating a data frame name bear_df with body mass data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>mass</th>\n",
       "      <th>charged</th>\n",
       "      <th>polar</th>\n",
       "      <th>hydrophobic</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Ursus spelaeus</td>\n",
       "      <td>550.800</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Ursus arctos</td>\n",
       "      <td>203.500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Ursus thibetanus</td>\n",
       "      <td>99.714</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Melursus ursinus</td>\n",
       "      <td>100.030</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Ursus americanus</td>\n",
       "      <td>110.560</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Helarctos malayanus</td>\n",
       "      <td>47.020</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Ailuropoda melanoleuca</td>\n",
       "      <td>118.200</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Tremarctos ornatus</td>\n",
       "      <td>140.700</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Ursus maritimus</td>\n",
       "      <td>425.100</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  species     mass  charged  polar  hydrophobic\n",
       "0          Ursus spelaeus  550.800      NaN    NaN          NaN\n",
       "1            Ursus arctos  203.500      NaN    NaN          NaN\n",
       "2        Ursus thibetanus   99.714      NaN    NaN          NaN\n",
       "3        Melursus ursinus  100.030      NaN    NaN          NaN\n",
       "4        Ursus americanus  110.560      NaN    NaN          NaN\n",
       "5     Helarctos malayanus   47.020      NaN    NaN          NaN\n",
       "6  Ailuropoda melanoleuca  118.200      NaN    NaN          NaN\n",
       "7      Tremarctos ornatus  140.700      NaN    NaN          NaN\n",
       "8         Ursus maritimus  425.100      NaN    NaN          NaN"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bear_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "species_list = list(bear_df.species)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "importing the species list"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "## 5 ##\n",
    "## Write a for-loop that translates each sequence and also gets the proportion\n",
    "## of each aa type in that translated sequence and adds those data to dataframe\n",
    "# for key, value in cytb_seqs.items():\n",
    "#     aa_seq = nuc2aa_translate_function(value) # whichever function you prefer of #2 or #3\n",
    "#     # get proportions of each aa type\n",
    "#     set the value of each proportion in the dataframe (i.e., fill in empty cells in DF)\n",
    "\n",
    "# %matplotlib inline # uncomment for jupyter nb."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for key, value in cytb_seqs.items():\n",
    "    aa_seq = sequence_data_dict.translate(to_stop=True)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "## 6 ##\n",
    "## Plot a bar-chart of the mass with the x-axes labeled with species names.\n",
    "## What is the largest bear species? What else is interesting about this species?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x11baa95f8>"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZQAAAEFCAYAAADE/xFGAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAH4RJREFUeJzt3XucXeO9x/HP5GaEEYMJUpdEW7+kdb/FLc1QhFOkyumh\noi51Sav0nDqHEvSCuhTHCUWbICpV1aggGpeWBCkSlyJN/EjdSlOdyFVzT+b88Xt2ZhuZyZ5krdmz\n6/t+veY1e6+991q/9axnPb/1rL32s6oaGxsRERFZV53KHYCIiPxrUEIREZFMKKGIiEgmlFBERCQT\nSigiIpKJLuUOYG00NCzQpWkiIm1UV1dTlef81UMREZFMKKGIiEgmlFBERCQTSigiIpIJJRQREcmE\nEoqIiGRCCUVERDKhhCIiIplQQhERkUwooYiISCYqcuiVjxlzX3mWe8zg8ixXRKQDUg9FREQyoYQi\nIiKZUEIREZFMKKGIiEgmlFBERCQTSigiIpIJJRQREcmEEoqIiGRCCUVERDKhhCIiIplQQhERkUwo\noYiISCaUUEREJBNKKCIikgklFBERyUSu90MxsxeA+enpm8BlwCigEZgKnOnuK83sNOAMYDlwqbuP\nyzMuERHJXm4JxcyqgSp3ry+adj9wobtPMLObgcFm9jRwNrAHUA08ZWaPuvuSvGITEZHs5dlD2Rno\nbmaPpOVcAOwOTEyvjwcOAVYAk1ICWWJmM4CdgCk5xiYiIhnLM6EsBK4GRgKfJRJIlbs3ptcXAD2A\njYB5RZ8rTG9RbW13unTpvOp5Q3Yxt0ldXU2Zliwi0vHkmVBeA2akBPKamX1A9FAKaoC5xHcsNauZ\n3qI5cxZmHOraaWhYUO4QRERKlvdBcJ5XeZ0CXANgZr2InsgjZlafXj8MeBKYDAwws2oz6wH0I76w\nFxGRCpJnD+UWYJSZPUVc1XUKMAsYYWbdgOnAGHdfYWbDieTSCRjm7otzjEtERHJQ1djYuOZ3dTAN\nDQs+GvSY+8oTyDGDy7NcEZG1UFdXU5Xn/PXDRhERyYQSioiIZEIJRUREMqGEIiIimVBCERGRTCih\niIhIJpRQREQkE0ooIiKSCSUUERHJhBKKiIhkQglFREQyoYQiIiKZUEIREZFMKKGIiEgmlFBERCQT\nSigiIpIJJRQREcmEEoqIiGRCCUVERDKhhCIiIplQQhERkUwooYiISCaUUEREJBNdyh2AiMi/upWj\n3yzLcjsN6dO+y2vXpYmIyL8sJRQREcmEEoqIiGRCCUVERDKhhCIiIplQQhERkUzketmwmfUEngcO\nBpYDo4BGYCpwpruvNLPTgDPS65e6+7g8YxIRkXzk1kMxs67Az4BFadK1wIXuPgCoAgab2RbA2cB+\nwCDgcjNbL6+YREQkP3n2UK4GbgbOT893Byamx+OBQ4AVwCR3XwIsMbMZwE7AlNZmXFvbnS5dOq96\n3pBt3CWrq6sp05JFpJK8X6bltncblUtCMbOTgAZ3f9jMCgmlyt0b0+MFQA9gI2Be0UcL01s1Z87C\nDKNdew0NC8odgohIi5q3UXknmLx6KKcAjWZ2ELAL8AugZ9HrNcBcYH563Hy6iIhUmFwSirt/ofDY\nzCYAQ4GfmFm9u08ADgMeByYDl5lZNbAe0I/4wl5ERCpMew4OeQ4wwsy6AdOBMe6+wsyGA08SFwgM\nc/fF7RiTiIhkJPeE4u71RU8Hrub1EcCIvOMQEZF86YeNIiKSCSUUERHJhBKKiIhkQglFREQyoYQi\nIiKZUEIREZFMKKGIiEgmlFBERCQTJf2w0cz6AIcDnwVWAjOAB9z97RxjExGRCtJqQjGzLYHrgG2B\nSUQiWQb0Ae42s7eAc9z93XzDFBGRjm5NPZQrgB+6+7TVvWhmOwOXAydkHZiIiFSWVhOKu5+4htdf\nQslEREQo/TuUvYD9gRuAccCuwFB3vyfH2EREpIKUepXXcOA54BhgIbAb8L28ghIRkcpTakLp5O5P\nAF8C7nH3v9K+91IREZEOrtSEstDMzgEOBMaZ2XeI+7+LiIgApSeU44ENgKPdfQ7QC/hablGJiEjF\nKfW0VQMw1t1fNrOvEYloRX5hiYhIpSm1hzIaOMbM+gM/BOYDt+cWlYiIVJxSE0ofd78YOBoY6e6X\nALX5hSUiIpWm1ITSxcw2A74MPGhmWwDd8wtLREQqTakJ5SfAs8CD7j4VeAL4UW5RiYhIxSnpS3l3\nvxO4s2hSP6BbLhGJiEhFKnXolaOBi4ENgSqgM3HKqy6/0EREpJKUesrrKuA/genEb1JuA36dV1Ai\nIlJ5Sk0oc9z9ceAZoIe7/wDYJ7eoRESk4pSaUBaZ2fZED6XezLoBPfILS0REKk2pCeVC4FJi6Pov\nAu8DY/MKSkREKk+pV3lNBCamp3uaWW0a00tERAQo/SqvAcSX8rVF03D3A1v5TGdgBGBAIzAUWAyM\nSs+nAme6+0ozOw04A1gOXOru49ZqbUREpGxKHRxyFDGG19ttmPcRAO6+n5nVA5cRlxxf6O4TzOxm\nYLCZPQ2cDewBVANPmdmj7r6kDcsSEZEyKzWhvOfuv2jLjN19rJkVehrbAnOBg2g6dTYeOIQYtXhS\nSiBLzGwGsBMwpS3LExGR8io1oQw3s9HAY8RpKQDWlGTcfbmZ3Q4cRdw++GB3b0wvLyCuFNsImFf0\nscL0FtXWdqdLl86rnjeUuBJZq6urKdOSRaSSvF+m5bZ3G1VqQvlW+j+gaFojsMZei7ufaGbnEWOB\nrV/0Ug3Ra5mfHjef3qI5cxaWEHL+Ghp000oR6biat1F5J5hSE8qW7t6vLTM2sxOArdz9cmAhsBJ4\nzszq3X0CcBjwODAZuMzMqoH1iHHCprZlWSIiUn6lJpQnzexw4CF3X77Gd4ffAreZ2RNAV5qGbhmR\nfhg5HRjj7ivMbDjwJPG7mGHuvrhNayEiImVX1djYuMY3mdlMYHPiNBfE1VqN7t655U/lp6FhwUeD\nHnNfOcKAYwaXZ7kiUlFWjn6zLMvtNKTPR57X1dVU5bm8Un/YuGWeQYiISOVrdegVM7vczDZu5fVN\nzOzK7MMSEZFKs6Yeyt3AWDP7G3GXxneJy4a3BQ4EehHfjYiIyCdcqwnF3V8kRhc+ADgSOJy4Wusv\nwM/c/bH8QxQRkUpQ6ncojxOX+IqIiKxWqYNDDiKGr9+EuMILAHffLqe4RESkwpT6O5Trge8SPzhc\n83XGIiLyiVNqQpmlIeVFRKQ1bfml/LXAQ8Q9TQBw9ydyiUpERCpOqQllr/R/16JpjcSlwyIiIiVf\n5XVA3oGIiEhlK/Uqr/2B/wE2JK7y6gxs6+698wtNREQqSatDrxQZCYwlEtBPgdeBe/MKSkREKk+p\nCWWRu98GTADmAKcBA/MKSkREKk+pCWWxmW0COLB3uo3vBvmFJSIilabUhHIt8GvgAeDrZvZn4Lnc\nohIRkYpTUkJx998Ah7j7AmB3YAhwQp6BiYhIZSkpoZhZLfBzM3sMqAbOAnrkGZiIiFSWUk95jQCm\nAJsCC4CZwOi8ghIRkcpTakLp4+4/B1a6+1J3HwZslWNcIiJSYUpNKMvNrAdppGEz+yxxoy0RERGg\n9LG8vk/8BmVrMxsL7AOckldQIiJSeUrtoTxP/DL+TWAb4LfE1V4iIiJA6T2U3wEvA8X3RKlq4b0i\nIvIJVGpCwd2/kWcgIiJS2UpNKGPN7FTgMWB5YaK7v5NLVCIiUnFKTSg9gO8Bs4qmNQLbZR6RiIhU\npFITytFAT3dflGcwIiJSuUq9yusNoDbPQEREpLKV2kNpBKaZ2VRgaWGiu+ue8iIiApSeUC5ry0zN\nrCtwK9AbWA+4FJgGjCKS01TgTHdfaWanAWcQX/Zf6u7jVjdPERHp2EpKKO4+sY3zHQJ84O4npBtz\n/Sn9XejuE8zsZmCwmT0NnA3sQYxi/JSZPeruS9q4vA5n5pjjy7LcLY/5ZVmWKyJS8u9Q2ug3wJj0\nuIrofewOFBLTeOAQYAUwKSWQJWY2A9iJGNlYREQqSC4Jxd0/BDCzGiKxXAhcnW4dDDEEfg9gI2Be\n0UcL01tVW9udLl06r3rekE3YbVZXV9PiazPbMY5ircUkIuXxfpmW297tQV49FMxsa2L8rxvd/U4z\nu6ro5RpgLjA/PW4+vVVz5izMMtS11tCwoNwhfExHjElEyqN5e5B3gin1suE2MbPNgUeA89z91jT5\nRTOrT48PA54EJgMDzKw6DY/fj/jCXkREKkxePZQLiN+tXGRmF6Vp3wGGm1k3YDowxt1XmNlwIrl0\nAoa5++KcYhIRkRzl9R3Kd4gE0tzA1bx3BHGLYRERqWC5nPISEZFPHiUUERHJhBKKiIhkQglFREQy\noYQiIiKZUEIREZFMKKGIiEgmlFBERCQTSigiIpIJJRQREcmEEoqIiGRCCUVERDKhhCIiIplQQhER\nkUwooYiISCaUUEREJBNKKCIikom8bgEsIlIWr40vz13Etz+suizL7UjUQxERkUwooYiISCaUUERE\nJBNKKCIikgklFBERyYQSioiIZEIJRUREMqHfoYjIWrvkqbllWe5F+29cluVK69RDERGRTCihiIhI\nJnTKS2Q1TnryjnZf5qgBJ7T7MkWypB6KiIhkItceipn1B65093oz+wwwCmgEpgJnuvtKMzsNOANY\nDlzq7uPyjElERPKRWw/FzM4FRgKFITivBS509wFAFTDYzLYAzgb2AwYBl5vZennFJCIi+cmzh/IX\n4CtA4WT07sDE9Hg8cAiwApjk7kuAJWY2A9gJmNLajGtru9OlS+dVzxuyjbtkdXU1Lb42sx3jKNZa\nTHeMGtSOkTQ54aSHW3zt5HsPbcdImtx21ENlWW5rWtt2HVd5LhturaxeozzD17cW0/vtGEex9q5T\nuSUUd7/HzHoXTapy98b0eAHQA9gImFf0nsL0Vs2ZszCrMNdJQ8OCcofwMYqpNIqpsnXEsqqEmPJO\nMO35pfzKosc1xKHN/PS4+XQREakw7ZlQXjSz+vT4MOBJYDIwwMyqzawH0I/4wl5ERCpMe/4O5Rxg\nhJl1A6YDY9x9hZkNJ5JLJ2CYu5fnBKiIiKyTXBOKu78F7J0evwYMXM17RgAj8oxDRETypx82iohI\nJpRQREQkE0ooIiKSCSUUERHJhBKKiIhkQglFREQyoYQiIiKZUEIREZFMKKGIiEgmlFBERCQTSigi\nIpIJJRQREcmEEoqIiGRCCUVERDLRnvdDEZF18I2JE8uy3FsGfuyuEyKrpR6KiIhkQglFREQyoYQi\nIiKZUEIREZFMKKGIiEgmlFBERCQTSigiIpIJJRQREcmEEoqIiGRCCUVERDKhhCIiIplQQhERkUwo\noYiISCaUUEREJBMdYvh6M+sE3AjsDCwBTnX3GeWNSkRE2qKj9FC+DFS7+z7A94BryhyPiIi0UUdJ\nKPsDDwG4+zPAHuUNR0RE2qqqsbGx3DFgZiOBe9x9fHr+DrCduy8vb2QiIlKqjtJDmQ/UFD3vpGQi\nIlJZOkpCmQT8G4CZ7Q28Ut5wRESkrTrEVV7AvcDBZvZHoAo4uczxiIhIG3WI71BERKTydZRTXiIi\nUuGUUEREJBNKKCIikomyJBQzqzezu5pNu8LMTipHPK0pxGpmvc3smTQtt1jXtWzM7Ntrudxvp/8n\nmdkVq4mp0cyON7NRZnZomv6ymY1qZZ7/a2Y/XZt4WplnbnUnrfuRWS7bzCaYWd9WXj/UzE5v5fVN\nzOxrRc/PNbOZZladnl9nZtuY2Q/MbGhL82mr4vpewnv7mtmEFl67JpXBq2b2Tnr8m7WIZ1W9W1ep\nrG42s14tvN7u7ZOZVZvZqRnM5y4z65bqxBFp2nVmts26R7lmHeUqL8nOhcANOXzuVeDfgbkAZrYj\nsMEa5tkHqF6LWMrC3UeVYZkPreEtOwFHAnem50OAu4BjgVHu/p8AZpZbjOvC3c+BSNZAX3f/Xnkj\nWuVA4Drgb+UOJNkCOBUYuS4zcfdjAczsQKAv8EChjrSHslzlZWb1wNDCyqdpVxCN1lvAlcBS4OdE\noRxAJL973P3KdDQ01N1fTUdlWwBXAHcDPYDuwDB3f6Ro/nXAr4leWTUwlGgcfwPMBLYCxrv7MDPb\nOi17/fT3d+A7wF3uvreZ3ZviWg5sC/yZGNxyKLAjcenzfe4+xMzmAse5+3gz+yXwKWAw4CmOrsAl\n7n5FUdl8E1gAbA1smR6PIBqR/kBn4H+A0wBL6zQGeA24hGhwTgSeAnql9esE/DPFuxiYA0xx97PN\nbBjwfaIyTwZOSuVfB9yU5vsg8Hpa1xqioQN43t2PNrNbgK8CjcCzwKHA+2m5VwNfTOXSCdgP+Dzw\nY2LHXgm8B2wKzAA+A4xK8YwF5rp7vZm9B3yYtvc84G1gYYrpZZrqTm/gdmAZcFRaj6uBb6XY709/\n16SynAx0K1qnG1K8S4HtiHrTO2239YmG6Kn0vg9TmValv85pHc919xvM7G1gVppvYyr3fYEX03bo\nl+bxIvBGirErUecOBR4h6sxM4M1UbkuBnqmcd02f3ZrY1temeN4i9qer0/zWI/ahfwATgINS+byb\n5rMD8Ln0ntfTeuLuu5jZ7akcSeW+K3BRiu9vwAfAIGBj4OH0enWavluKYdv0twHwXNome6Rl/ZOo\nxzOB2cS+1Zi2yRvEdu5P1IntgXqinnQFpqay2CMt7w3iQKYhrd8rKc66tJybifp1VPrs48SBUk+i\nDo5N67IwbcdN03a9P8W/A9FWNKT5bZXWaRlRD+toqmMQ9acXsCitG0RdmkMMhDsrxfLbFOcMog17\nLG2LPYHl7r69md2Ttlun9Lk9UzzF7eUlRB15nmgHvw18l2ibjk3rvllar58CR6cyPZGoc3e5+94A\nqYd6LFH/rknruBA4xt0XsBod7TuUQnardvcB7n4HcDzwNWAA6ei4BZ8mCuoI4Dg+3vvai6hwhwFn\n0nR03ZtoQPcEDjSzwg4w3N3ricZk52bz2oTYcc8iCvkGYgPvQjRKnwIGmtnGLcQ6kNjYn0/xdmv2\nenfgGXcflOLelaayWejuGxENQz8igW0HbObulxFJ7kfAGUQiuiWt7+eJSjMfuCkNxDndzLqkz812\n92+lZSwjdqqjgOKjm7HETvVHonG7BdjDzPYhynxLoDbF9HWi0Z2c3t8DGJLKdDGxE0M0dH2IHlIt\nsS0eTNOb24wo65lEIziQSHjVReVTDYwndsjjiZ1tPFF3tgamEQcfxxGN9a3Ezrc58Ku0Tp3Teh4N\n7A2cCzyT3rccOA+4itjOfVJ5TSES6OlE3TjfzI4hGubj0/ptnNbbgQ3d/aC0npNSnCcB56dymJzW\n72GiselDJKLRxMHFPKJReAPYnainC4jkXKxrKpOxRILelGjs/kjsI/sTdfBSYjy9F4FxRNLcJvVE\nDyQazh5F618LdHX3A1Is76fynZ/+b0o0roNSHB8QB3zPpnU5N22zl1L8nYDpaX0hBom9PpXxfxGN\n/5eIA5+fEElmEHAw0ci+kJZ9MU0HPd8k9p/lKZZJRNJ8m0ikI4j9tYE4SFmUXt+IqA93p235XFr/\nHdM8/0I04JcCK4ge5LQUa3Edg0gio4kDPkvvHUsc8PanKWmMTvPckaj/hwEnEAeNhVNzBxKJdTPg\nr0VlW9xekmK6ArjT3QuJrWCRux8K3AP8m7sfkd57LC37ciqLwv5W29Iby5VQFvHxBmPDNB1ihys4\nnljhh4kdsrkqAHf/M/AzolG4kY+v23iiQt1HNLiFHe8ld5/t7iuIym7ERr0g9YSGEA18sRXAKcSO\n1UjstDsSlWcy8E6a1ruFWO8nehEvEhuqc9F7CmWwZ+rR/C9R6QrT/5L+G7FDXEE0Jk80W1Y/4mgH\nd59C7Ky/JI6GjzCziTQdWTf3grs3Ekcsxet+F9EYzAeeJHacjYgjxvWInfSDNK35AJ+LgOHpO5du\nRPkA/DUd7TxI7Ni/IHqkzRtG0vp8lWgcegJ/AI4hGsyW6s4pwCFE3ZlFNC4QR219iUZ8WyKxHE0k\nmk7AK+6+3N3/mea9J7GN/0kkxPOII9qHidNRu6T5XEzsnF2IntiGaZ1+mOIeTTQQS4vinE1sz6XA\nH919qbsfR/SyaoF9gNtSmfUn6uQSon4tIhqXI9KyCgdShe26lNiOA4htMpto9CYCje6+jGisf0wk\nl43Tcl4gkmHfVKa3pDKqSuu5BfAPM/scccD1fpr3pkRd+xMf3c5vp//ziQQEsR9NS+XSNa3n0USd\nu5Y4ql4BfDatx1/TcvYkEsWwtK1+lcqve5rPbKIH0JeoM1PdfRFRZ58rKvvtie32KSKJdiV6dd3S\nsnoRB1SfSzG+RtTLV4mDiN8TSfimNJ/107KLh416N/1/luiN3UmMCnJe6tVvVVRGC9P2WAzMcvel\nxEFCoS27mTgIeittg8Lniuv8mryQ/s8lyh6irFZ3arpQh35MlEVhf1vW0szLlVCmA7ua2ZYQX0gB\nX6BpZVem6esRR3THEY3MSWZWOGWzZXrvbum9OwI17v4l4kj8+mbLrAdmuvshxJHFj9P0fmbW3cw6\nEzvrNKLCnJeOpk8jyqkuLacnsXOeSBwprSAKfgaxkTYnejCdiYq7jKjYxbF+BdjY3XsSp1YuaFY2\n+wLL3P34tB6FHXxV2RCVc1Aqm8HARalsGlPZTCcaNMysP7AN0eP4M/AZdx9I9Hz2TfMrTiwtnQd9\ni2iwziAaxh3SOj9NJKo6d9+YOAJ7IM2nithe/YkREL6f1qewvMKy6lNZfZ3YUb+aPtcrrUNPIpl9\nM72/N5EMnNhBi8tnMdFo/juxAz6U3gtNded0IvGPIhrB/YidfRRwdrMyWD+t5/NE0utO2pZEIu1N\nNNDPpnndRJyCeyZ97utEw3I/sb2a97QbiTrXBdjLzLqa2RQigdel+b5K1LVbibK/I72/juht/5po\nzE5P67dLiqWWqIePpvdsTNSN/VO5diVOtfw30aMsbLPdiWT1LrFvHk+c49+AaNSnEaeTLiKS2zIi\nAexG9JiPJPab1R2wFFQRjfe8tNzJRPm/QfRGfpnm/XLRZ85Jz18mndZx9/8g9ofi5RXK9NNA39SW\nfIHYV2qIevJWWud3iZ7O3cS+OiutxzjiQG0W0ZOpS/NvJHoJ303lNDiV6YY01fdC+/Tp9L83Ue/P\np+l09llFMa8sin0JTWctdoZVp+zPIvaBrdLyN0vvWd3B10pW37639h3HYqCnmXVOZ1f6pOlDiO/r\nDiDajxYvIilLQnH3+cTGeNDMniI22vXNb6rl7kuIneQZ4lznI0QjMBy40cwepuno/nWg3syeIL4X\nubjZYl8CTk29jp8Al6fpS9P7nyW+93iJ2Lm+n47ibyJOgd1GHKmMI46+Rqc4VhCN3r1ERZmX/mYQ\nlex64Aozm0VTef8B2Dd9v/IEcfRXXDY/Ak4xs3lEY/13mo7AC36bljOfOJr6fSqbV4kG/UjiqOOU\ntA5vEt3uemClmT1GnDZ7Ns1vmpmNZs3eJCr0zURX+Dl3n0D0MD4wsw+JI8iniIZnZ6KBeSetx2Ti\nKL/5FTYvETvsXenzS9PjmcRO+1Nix76DaJB7pc/8N3Bds7ozktj5jiCSQ+G7iJWkukM0WIcSjX01\n0dCeTOwstzaLbWl6b1/i1OFi4uiW9LgHcWpsByKx/hex495N1I/RRAM1KJVTLc1OG7j7n4jG9Eqi\nl7dherwJcDjwA6JOnZ7i/CqxnT9FNFTfIBqkJSnOPYgDiPdSuR5KJNV57j6O2I6bEPvW08SpxMOI\nA49TiYOGt9Pr04j9cC5Rx/9OHJQVTgt+mFbjwbS+c2jazi1e4UbUv1OJ/e9N4qDjP4j6dW8h3rQO\nBQ+k9+yV1m+FxZBNOxPJflW9cvdXiIOEbqlM+xP77wIiaZ2cPtMzrdvpqYwvIPa33xP1Y1+izRib\nyumg9HxfIlE8l8qi8F1ZoY4dTNM+/x7Rq7uR2B+WEUl+Zop5LtDJzK4Efgdsktqqwun2WUQCnEds\ni3+w+tPCBa8Ag82stVNZH+Huf08xTSHOfhT2qcnASDP7A3Ha7RctzeMTPfSKmfWm6EsokU+ydDry\nrhKuPCu8vzvRM+vv7qs7SpZPmI72pbyIVAAz25fo3V6pZCIFn+geioiIZEc9FBERyYQSioiIZEIJ\nRUREMqGEIpIjM+tlZr8rdxwi7UFfyouISCY02rAIYGZbEb/M3oD4AeTZxA8r7yN+YQ1wiru/aGaf\nIX4suikxWN5Zafq2xA9ge6bppxI/PJ3g7r3NbHNieKCt0zLOd/ffm9kXibHBCoNHHufuhR/JiVQM\nnfISCd8Axrn7HsTAhfun6bPdfVdi5IXb07TbidGEdyN+XV24d8aNxIjYOxC/bL+w2TL+D7jV3Xcn\nRjL4mZnVpPcNTct+gKZhXUQqik55iQBmth8xnM0fiCFE7iGGsdnH3Wem98wmxnR6g6aB9SCGNtmJ\nGKqkVxo+pzDf3jT1UGbRNFggxNAnRxLj1J1DDO1xn7s/msc6iuRNp7xEAHeflEbOPZwYK+qk9FLx\nyLGdiLHjFrv7LoWJ6XTZbIpGYTWzKmLE54VFn+8MHOjus9N7egHvu/ufzOyBtOyrzGxMuqWASEXR\nKS8RwMyuAk5w99uJmxIVTjsV7oB3FDDd3d8GXjezIWn6wTTdOuAJmu4rcRBxw6NijxE30CIlr5eB\n7mb2LDFS9nXE7Qp0yksqkk55iQDpLp13EkObryBG+r2KGG23LzFy7snu/prFPeJvJk5ZLQW+6e5T\n0jxGErcwKHwpv5CmU169iCSzDTEy8LnpTp5fJO7/sZwY5Xaou09tp1UXyYwSikgLzOwtoN7d3ypv\nJCKVQae8REQkE+qhiIhIJtRDERGRTCihiIhIJpRQREQkE0ooIiKSCSUUERHJxP8DeHy9gHHAsKsA\nAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x11ba59898>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.barplot(x='species', y='mass', data=bear_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "using the seaborn package for data visualization to create the plot  \n",
    "specifying the x and y axes and the dataframe\n",
    "\n",
    "The biggest species is the Ursus spelaeus. According to Wikipedia, this species lived during the Pleistocene and became extinct about 24,000 years ago (https://en.wikipedia.org/wiki/Cave_bear)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "## 7 ##\n",
    "## Plot a visualization of the proportions for amino-acid type for the bear species\n",
    "## What does this show about cytochrome-b for the bears?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "    <div class=\"bk-root\">\n",
       "        <a href=\"http://bokeh.pydata.org\" target=\"_blank\" class=\"bk-logo bk-logo-small bk-logo-notebook\"></a>\n",
       "        <span id=\"5a808f07-91f9-4a61-992d-9e7ff9df4b4c\">Loading BokehJS ...</span>\n",
       "    </div>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": [
       "\n",
       "(function(global) {\n",
       "  function now() {\n",
       "    return new Date();\n",
       "  }\n",
       "\n",
       "  var force = true;\n",
       "\n",
       "  if (typeof (window._bokeh_onload_callbacks) === \"undefined\" || force === true) {\n",
       "    window._bokeh_onload_callbacks = [];\n",
       "    window._bokeh_is_loading = undefined;\n",
       "  }\n",
       "\n",
       "\n",
       "  \n",
       "  if (typeof (window._bokeh_timeout) === \"undefined\" || force === true) {\n",
       "    window._bokeh_timeout = Date.now() + 5000;\n",
       "    window._bokeh_failed_load = false;\n",
       "  }\n",
       "\n",
       "  var NB_LOAD_WARNING = {'data': {'text/html':\n",
       "     \"<div style='background-color: #fdd'>\\n\"+\n",
       "     \"<p>\\n\"+\n",
       "     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n",
       "     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n",
       "     \"</p>\\n\"+\n",
       "     \"<ul>\\n\"+\n",
       "     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n",
       "     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n",
       "     \"</ul>\\n\"+\n",
       "     \"<code>\\n\"+\n",
       "     \"from bokeh.resources import INLINE\\n\"+\n",
       "     \"output_notebook(resources=INLINE)\\n\"+\n",
       "     \"</code>\\n\"+\n",
       "     \"</div>\"}};\n",
       "\n",
       "  function display_loaded() {\n",
       "    if (window.Bokeh !== undefined) {\n",
       "      document.getElementById(\"5a808f07-91f9-4a61-992d-9e7ff9df4b4c\").textContent = \"BokehJS successfully loaded.\";\n",
       "    } else if (Date.now() < window._bokeh_timeout) {\n",
       "      setTimeout(display_loaded, 100)\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function run_callbacks() {\n",
       "    window._bokeh_onload_callbacks.forEach(function(callback) { callback() });\n",
       "    delete window._bokeh_onload_callbacks\n",
       "    console.info(\"Bokeh: all callbacks have finished\");\n",
       "  }\n",
       "\n",
       "  function load_libs(js_urls, callback) {\n",
       "    window._bokeh_onload_callbacks.push(callback);\n",
       "    if (window._bokeh_is_loading > 0) {\n",
       "      console.log(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n",
       "      return null;\n",
       "    }\n",
       "    if (js_urls == null || js_urls.length === 0) {\n",
       "      run_callbacks();\n",
       "      return null;\n",
       "    }\n",
       "    console.log(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n",
       "    window._bokeh_is_loading = js_urls.length;\n",
       "    for (var i = 0; i < js_urls.length; i++) {\n",
       "      var url = js_urls[i];\n",
       "      var s = document.createElement('script');\n",
       "      s.src = url;\n",
       "      s.async = false;\n",
       "      s.onreadystatechange = s.onload = function() {\n",
       "        window._bokeh_is_loading--;\n",
       "        if (window._bokeh_is_loading === 0) {\n",
       "          console.log(\"Bokeh: all BokehJS libraries loaded\");\n",
       "          run_callbacks()\n",
       "        }\n",
       "      };\n",
       "      s.onerror = function() {\n",
       "        console.warn(\"failed to load library \" + url);\n",
       "      };\n",
       "      console.log(\"Bokeh: injecting script tag for BokehJS library: \", url);\n",
       "      document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "    }\n",
       "  };var element = document.getElementById(\"5a808f07-91f9-4a61-992d-9e7ff9df4b4c\");\n",
       "  if (element == null) {\n",
       "    console.log(\"Bokeh: ERROR: autoload.js configured with elementid '5a808f07-91f9-4a61-992d-9e7ff9df4b4c' but no matching script tag was found. \")\n",
       "    return false;\n",
       "  }\n",
       "\n",
       "  var js_urls = [\"https://cdn.pydata.org/bokeh/release/bokeh-0.12.4.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-widgets-0.12.4.min.js\"];\n",
       "\n",
       "  var inline_js = [\n",
       "    function(Bokeh) {\n",
       "      Bokeh.set_log_level(\"info\");\n",
       "    },\n",
       "    \n",
       "    function(Bokeh) {\n",
       "      \n",
       "      document.getElementById(\"5a808f07-91f9-4a61-992d-9e7ff9df4b4c\").textContent = \"BokehJS is loading...\";\n",
       "    },\n",
       "    function(Bokeh) {\n",
       "      console.log(\"Bokeh: injecting CSS: https://cdn.pydata.org/bokeh/release/bokeh-0.12.4.min.css\");\n",
       "      Bokeh.embed.inject_css(\"https://cdn.pydata.org/bokeh/release/bokeh-0.12.4.min.css\");\n",
       "      console.log(\"Bokeh: injecting CSS: https://cdn.pydata.org/bokeh/release/bokeh-widgets-0.12.4.min.css\");\n",
       "      Bokeh.embed.inject_css(\"https://cdn.pydata.org/bokeh/release/bokeh-widgets-0.12.4.min.css\");\n",
       "    }\n",
       "  ];\n",
       "\n",
       "  function run_inline_js() {\n",
       "    \n",
       "    if ((window.Bokeh !== undefined) || (force === true)) {\n",
       "      for (var i = 0; i < inline_js.length; i++) {\n",
       "        inline_js[i](window.Bokeh);\n",
       "      }if (force === true) {\n",
       "        display_loaded();\n",
       "      }} else if (Date.now() < window._bokeh_timeout) {\n",
       "      setTimeout(run_inline_js, 100);\n",
       "    } else if (!window._bokeh_failed_load) {\n",
       "      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n",
       "      window._bokeh_failed_load = true;\n",
       "    } else if (force !== true) {\n",
       "      var cell = $(document.getElementById(\"5a808f07-91f9-4a61-992d-9e7ff9df4b4c\")).parents('.cell').data().cell;\n",
       "      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n",
       "    }\n",
       "\n",
       "  }\n",
       "\n",
       "  if (window._bokeh_is_loading === 0) {\n",
       "    console.log(\"Bokeh: BokehJS loaded, going straight to plotting\");\n",
       "    run_inline_js();\n",
       "  } else {\n",
       "    load_libs(js_urls, function() {\n",
       "      console.log(\"Bokeh: BokehJS plotting callback run at\", now());\n",
       "      run_inline_js();\n",
       "    });\n",
       "  }\n",
       "}(this));"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from bokeh.io import push_notebook, show, output_notebook\n",
    "from bokeh.charts import Histogram\n",
    "output_notebook()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'bears_df' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-29-4f8abf9bc03e>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mhist\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mHistogram\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbears_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mbears_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mspecies\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'OL'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalues\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"species\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcolor\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"charged\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mshow\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhist\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'bears_df' is not defined"
     ]
    }
   ],
   "source": [
    "hist = Histogram(bears_df[bears_df.species == 'OL'], values=\"species\", color=\"charged\")\n",
    "show(hist)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "## 8 ##\n",
    "## Save the new dataframe to a file called \"bears_mass_cytb.csv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df.to_csv(bears_mass_cytb)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "## BONUS ##\n",
    "## What else can we do with this dataset in Python? \n",
    "## Add functions or anything that might be interesting and fun. (optional)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
